### ISSUE TITLE
lost values in map(compose(memorystream)).flatten() in xstream 6.0+

### ISSUE THREAD

**Issue by svachalek** [staltz/xstream#144]
In this demo program, which is essentially a simple flatmap:

```javascript
import xs from 'xstream';

const source$ = xs.of('a', 'b');
const integrate1$ = xs.of('1').take(1).remember();
const integrate2$ = xs.of('2').take(1).remember();

const out$ = source$
  .map(
    source => xs
      .combine(integrate1$, integrate2$)
      .map(([integrate1, integrate2]) => {
        return [source, integrate1, integrate2].join(' ');
      })
      .debug('combine')
  )
  .debug(s => console.log('map'))
  .flatten()
  .debug('flatten');

out$.addListener({
  next: v => console.log('next', v),
  error: console.error,
  complete: _ => console.log('complete')
});
```

Using xstream 5.3.6 I get the following output as expected:

```
map
combine: a 1 2
flatten: a 1 2
next a 1 2
map
combine: b 1 2
flatten: b 1 2
next b 1 2
complete
```

Using xstream 6.0 and 7.0 I get the following:

```
map
combine: a 1 2
flatten: a 1 2
next a 1 2
map
```

It seems that the inner combine does not see the remembered values on the source streams, as I would expect.

**Comment by staltz** [comment]
v5.3.6 was bugged and the current behavior is correct, due to how xstream handles start/stop of streams.

If you had written `const source$ = xs.periodic(100).take(2).map(i => ['a', 'b'][i]);`, it would work as expected. Previously, with source$ emitting synchronously, it would flatmap values to another synchronous stream, the combine of those two integrate$. When the second value `b` was emitted on `source$`, it would attempt to synchronously switch from the previous inner to the current inner. The current inner would only "stop" and restart if it was unused for a full iteration in the event loop.

This sync-start/async-stop dynamics is explained in the README.

In short, if you want to use "flatMap" and not "switchMap", use `map` + `flattenConcurrently`, not `map` + `flatten`. Finally, if all your streams are synchronous, it's not the best use case for xstream. xstream is best for hot, asynchronous data sources. Yours were mainly cold, synchronous. You could use RxJS for that purpose, then convert to xstream using `xs.fromObservable`.

**Comment by svachalek** [comment]
Actually this is a contrived example which was an attempt to simplify something I was seeing in a more complex scenario. In reality everything I have is async, but one of the streams was a fromPromise() which probably exhibits the same stop() behavior. (I've since removed that for other reasons, but it looks like now that it's gone the code in question is indeed working fine on 7.0.0.)

Thanks for the detailed explanation, it really helps how I think about it.
